src/main/java/org/eln2/mc/common/blocks/CellTileEntity.kt

```diff
@@ -0,0 +1,736 @@
+    // null before setPlacedBy
+    private var _connectPredicate : ((dir : Direction) -> Boolean)? = null
+    // used to fetch the cell factory from registry
+    // null before setPlacedBy
+    private var _cellId : ResourceLocation? = null
+
+    val connectionPredicate get() = _connectPredicate!!
+    val cellId get() = _cellId!!
+
+    // this returns the adjacent tiles or fetches them then caches them.
+    // this does not affect performance if neighbours are not updated!
+    // see _adjacentCellUpdateRequired above
+    val adjacent get() = getAdjacentCellTilesFast()
+
+    /**
+     * Called by the block when one of our neighbours is updated.
+     * We use this to invalidate the adjacency cache. Next time we use it, we will query the world for the changes.
+     * @see _adjacentCellUpdateRequired
+    */
+    @In(Side.LogicalServer)
+    fun neighbourUpdated(pos : BlockPos) {
+        _adjacentCellUpdateRequired = true
+    }
+
+    /**
+     * Called by the block when it is placed.
+     * For now, we are only processing this for the server.
+     * It will add our cell to an existing circuit, or join multiple existing circuits, or create a new one
+     * with ourselves.
+    */
+    @In(Side.LogicalServer)
+    fun setPlacedBy(
+        level : Level,
+        position : BlockPos,
+        blockState : BlockState,
+        entity : LivingEntity?,
+        itemStack : ItemStack,
+        cellProvider: CellProvider
+    ) {
+        if(level.isClientSide){
+            return
+        }
+
+        _cellId = cellProvider.registryName
+
+        // map the cell ID to a factory
+        _cell = cellProvider.create(pos)
+        cell.setId(cellId)
+
+        _connectPredicate = {
+            cellProvider.connectionPredicate(it)
+        }
+
+        _manager = CellGraphManager.getFor(level as ServerLevel)
+
+        registerIntoCircuit()
+
+        // call place handler
+        cell.setPlaced(getAdjacentCellsFast())
+
+        setChanged()
+    }
+
+    /**
+     * Called by the block when it is broken.
+     * For now, we are only doing processing for the server.
+    */
+    @In(Side.LogicalServer)
+    fun setDestroyed() {
+        cell.destroy()
+
+        val adjacent = getAdjacentCellTilesFast()
+        _logger.info("${this.pos} is performing a remove!")
+
+        if(adjacent.count() < 2){
+            // we cannot split a circuit because we are adjacent to only one component.
+
+            removeToOneCircuit()
+
+            if(adjacent.isEmpty()) {
+                // we are actually alone, we can destroy the graph.
+                graph.destroy()
+                _logger.info("Also destroyed our graph.")
+            } else {
+                removeOurselvesFromTheirCell(adjacent[0])
+                adjacent[0].updateCellConnectionsButExclude(cell)
+            }
+
+            return
+        }
+
+        // it is possible that our removal will split the graph into multiple pieces.
```

Comment:
This is not quite how we should go about this. In particular, the invalidation strategy is not needed since we already know that these cells need to be checked. Time for some impromptu graph theory:

At this point, there are anywhere from 2 to 6 neighbors to a cell. We'll pick one to be our starting cell and declare the rest to be destinations. The graph has not been cleaved into multiple graphs if and only if for all destinations we can find a path from the starting cell to that destination cell.

How should we go about doing this? The go-to algorithm for this is a breath-first search. In our case, we'll want the version in which we maintain a set of all of the cells we've seen and a stack of cells that still need to be visited. If we run out of cells to visit without finding any of the destinations, we've proved that the graph has been cleaved and everything in that set is a new graph. Rinse and repeat for the remaining neighboring cells until we've covered every one of them. 

If you want me to implement this, I totally can, but I wanted to give you a chance to try it.